Allow yield inside of try / catch and catch (#8413)
dotnet/csharplang: proposals/yield-in-try-catch.md

For example this is the `Dispose` method for the above iterator:

```csharp
[DebuggerHidden]
void IDisposable.Dispose()
{
    int num = <>1__state;
    if ((uint)(num - -4) > 1u && num != 1)
    {
        return;
    }
    try
    {
        if (num != -4 && num != 1)
        {
            return;
        }
        try
        {
        }
        finally
        {
            <>m__Finally2();
        }
    }
    finally
    {
        <>m__Finally1();
    }
}
```

This behavior is important to understand when considering the code generation for `try / catch` blocks.

### Code generation yield inside try with catch

The code generation for iterators that have `yield` inside `try` blocks must preserve the same exception semantics for the `catch` in both `MoveNext` and `Dispose`. To achieve this the compiler will take a similar approach to what it does for `finally` blocks.

For every `catch` block where the `try` has a nested `try / finally` with `yield`:

1. The contents of the `catch` block will be generated into a parameterless method on the iterator with a `void` return.

> 2. The contents of the `when` clause will be generated into a parameterless method on the iterator with a `bool` return.

Reviewer: Ah, I see that we don't need parameters for this extracted method after all. Makes sense :-)

Author: Yeah I realized it too once I started digging through a bunch of our generated code.

        }
    }
    catch (InvalidOperationException ex1) when (ex.Message.Contains("hello"))
    {
        Console.WriteLine("Catch1");
    }
    catch (Exception ex2) 
    {
        Console.WriteLine("Catch2");
    }
    finally
    {
        Console.WriteLine("Finally Outer");
    }
}
static T M<T>(T t) => t;
```

Will generate the following `Dispose` method:

```csharp
[DebuggerHidden]
void IDisposable.Dispose()
{
    int num = <>1__state;
    if ((uint)(num - -4) > 1u && num != 1)
    {
        return;
    }
    try
    {
        if (num != -4 && num != 1)
        {
            return;
        }
        try
        {
        }
        finally
        {
            <>m__Finally2();
        }
    }
    catch (InvalidOperationException ex) when (<>1__ex1 = ex, <>m__When1())
    {
        <>m__Catch1();
    }
    catch (Exception ex)
    {
        <>1__ex2 = ex;
        <>m__Catch1();
    }
    finally
    {
        <>m__Finally1();
    }
}
```

The `<>1__ex1 = ex` in the `when` clause is not legal but the IL generated for the `when` will conceptually have this behavior.

### Code generation yield inside try / catch in async iterators

> The code generation for `try / catch` blocks in async iterators will be largely the same as traditional iterators. The difference is that the return type of generated `catch` methods will be `ValueTask` instead of `void`.

Reviewer: > The code generation for `try / catch` blocks in async iterators will be largely the same as traditional iterators.

I didn't follow this. Async iterators are more like async state machines than iterator state machines. 
Instead of generating a MoveNext and a Dispose method with similar structure, we use a single method (MoveNext) for both normal and disposal execution. A flag indicates if we're in disposal and we use it to skip over non-disposal code as appropriate.

Here are some notes on [disposal in async iterators.](https://github.com/jcouv/roslyn/blob/d0a56dda8423ad4ddd1f100c66ab56c0f9a343c3/async-iterator.md#design) Happy to discuss offline.

Author: Hmm. Maybe i was looking at the wrong code gen here. Will take a look at this.

For example:

```csharp
foreach (var e in Iterator())
{
    Console.WriteLine(e);
}

IEnumerable<int> Iterator()
{
    try
    {
        yield return 1;
        throw new Exception("");
    }
    catch (Exception ex)
    {
        yield return 2;
    }
}
```

This code will output:

```cmd
1
2
```

The `yield` inside the `catch` block will go through the same rewriting as an `await` inside of a `catch` block. That will be observable when the `throw;` statement is used to rethrow an exception as it will reset the stack trace vs. preserving it (just as it is for an `async` method).

Detailed notes:

- The `yield` statement will be allowed in a `try / catch` block
- The `yield` statement will be allowed in a `catch` block

### Dispose and finally in iterators

A lesser known detail of iterators is that `finally` blocks can be executed as part of the `IDisposable.Dispose` implementation. The `Dispose` method has the same state machine structure as the generated `MoveNext` except it only has the parts necessary for executing the `finally` blocks. That allows `Dispose` to _resume_ the method from the last suspend and execute the `finally` that were _active_ at the last suspend point.

> The `Dispose` method _only_ executes the `finally` blocks and ignores all code inbetween them. For example consider this code sample:

Reviewer: Typo: inbetween doesn't appear to be a word
```suggestion
The `Dispose` method _only_ executes the `finally` blocks and ignores all code between them. For example consider this code sample:
```

```csharp
var e = M1(true).GetEnumerator();
e.MoveNext();
e.Dispose();

IEnumerable<int> M1(bool b)
{
    try
    {
        try
        {
            try
            {
                yield return 13;
            }
            finally
            {
                Console.WriteLine("inner finally");
                if (b)
                {
                    throw new Exception();
                }
            }
        }
        catch
        {
            Console.WriteLine("catch");
        }

        Console.WriteLine("after catch");
    }
    finally
    {
        Console.WriteLine("outer finally");
    }
}
```

This program will output:

```cmd
inner finally
outer finally
Unhandled exception. System.Exception: Exception of type 'System.Exception' was thrown
```

The `"after catch"` is not printed beacuse only the `finally` structure is mirrored in the `Dispose` method. The `catch`, like all other statements between `finally` is not included. Statements in between the `catch` and `finally` are not executed in `Dispose`. This may seem odd at first glance but is leaning into the specified behavior for iterator `Dispose`.

### Dispose and finally in async iterators

> The `DiposeAsync` behavior for async iterators mirrors that of traditional iterators in that it executes the `finally` blocks at the the point the state machine is suspended. However it does this by setting the `state` variable to a value that represents disposing and then calls `MoveNextAsync`. The implementation of `MoveNextAsync` will then execute only the `finally` blocks for the suspend point. The `DisposeAsync` method does not have a mirror copy of the `finally` blocks.

Reviewer: >it does this by setting the `state` variable [](http://example.com/codeflow?start=183&length=44)

It's not the "state" variable that indicates disposal mode, but a dedicated variable. It's called "disposeMode" in the speclet: [https://github.com/dotnet/roslyn/blob/main/docs/features/async-streams.md#detailed-design-for-async-iterator-methods](https://github.com/dotnet/roslyn/blob/main/docs/features/async-streams.md#detailed-design-for-async-iterator-methods)

### Code generation of yield inside try / catch in async iterators

The code generation of async iterators will change such that `catch` blocks do not observably execute during `DisposeAsync`. To achieve this all `catch` blocks visible from a `yield` will rethrow exceptions if the state machine is in a disposing state. For example consider the following code:

```csharp
var e = M(true).GetEnumerator();
await e.MoveNext();
await e.DisposeAsync();

async IAsyncEnumerable<int> M(bool b)
{
    try
    {
        try
        {
            yield return 13;
            await Task.Yield();
        }
        finally
        {
            Console.WriteLine("inner finally");
            if (b)
            {
                throw new Exception();
            }
        }
    }

>     catch

Reviewer: What if the catch has a `when` clause with side-effects? Would the side-effects be executed?

    {
        Console.WriteLine("catch");
    }
    finally
    {
        Console.WriteLine("outer finally");
    }
}
```

This will output:

```cmd
inner finally
outer finally
```

To achieve this the `catch` will be effectively rewritten as follows:

```csharp
catch (Exception ex)
{
    if (<>1__state == /* disposing state */)
    {
        throw;
    }
    Console.WriteLine("catch");
}
```

The generator will also need to modify any `when` clauses on `catch` blocks to ensure they don't execute during `DisposeAsync`. To do this a prefix will be generated that returns `false` when the state machine is in a disposing state. For example:

```csharp
// User code
catch (Exception ex) when (SomeMethod(ex))

// Generated code
catch (Exception ex) when (<>1__state == /* disposing state */ ? false : SomeMethod(ex))
```

### Code generation of yield inside catch

> The restrictions on the feature mean that `yield` inside a `catch` cannot be observed from a `finally` block. That means the code generation does not need to consider the impact on `Dispose` and can focus soley on `MoveNext`.

Reviewer: typo: soley -> solely
```suggestion
The restrictions on the feature mean that `yield` inside a `catch` cannot be observed from a `finally` block. That means the code generation does not need to consider the impact on `Dispose` and can focus solely on `MoveNext`.
```

This would add a bit of complexity to the feature and it only produces observable differences when a `finally` block throws an exception on the `Dispose` path. That is likely a rare case. Further it potentially increases the complexity for developers. The `Dispose` method at first glance is likely unintuitive in that it only mirrors `finally` blocks and no other statements but that is also a very simple rule to learn. Preserving `catch` and `finally` and discussing how the code flows between them is potentially more complex.

This also brings into question what happens when a `yield` occurs in a `catch` during `Dispose`. That cannot execute correctly as the state machine can't suspend during `Dispose`. To account for this the feature likely needs further restrictions like:

- The `yield` statement will be allowed in a `catch` block provided that:
  - The `try` block does not contain a `finally` block.
  - The `try` block does not have a nested `finally` block.

On the whole it seems simpler to not support `catch` in `Dispose` paths.

## Related Issues

Related Items:

> - https://github.com/dotnet/csharplang/discussions/765

Reviewer 2: Could be good to link to other duplicate items:
- https://github.com/dotnet/csharplang/discussions/6819
- https://github.com/dotnet/csharplang/discussions/8994
- https://github.com/dotnet/docs/issues/42112

}
```

> These restrictions exist in large part because it presented technical challenges for the native compiler and it wasn't a high enough priority item. That is no longer a blocker as the state machine in Rsolyn supports the types of transforms necessary to support this feature. There are still several semantic challenges to work through and this proposal has been written to address those.

Reviewer 3: ```suggestion
These restrictions exist in large part because it presented technical challenges for the native compiler and it wasn't a high enough priority item. That is no longer a blocker as the state machine in Roslyn supports the types of transforms necessary to support this feature. There are still several semantic challenges to work through and this proposal has been written to address those.
```

This behavior is important to understand when considering the code generation for `try / catch` blocks.

### Code generation of yield inside try with catch in iterators

The code generation for iterators that have `yield` inside a `try / catch` will preserve the `catch` semantics in the `Dispose` method. That is _only_ the contents of the `finally` blocks will execute even in the case exceptions come into play. For example:

```csharp
var e = M1(true).GetEnumerator();
e.MoveNext();
e.Dispose();

IEnumerable<int> M1(bool b)
{
    try
    {
        try
        {
            try
            {
                yield return 13;
            }
            finally
            {
                Console.WriteLine("inner finally");
                if (b)
                {
                    throw new Exception();
                }
            }
        }
        catch
        {
            Console.WriteLine("catch");
        }

        Console.WriteLine("after catch");
    }
    finally
    {
        Console.WriteLine("outer finally");
    }
}
```

This program will output:

```cmd
inner finally

> outer finally

Reviewer 3: Consider adding the exception to the output since it seems it will be thrown from the Dispose invocation, right?

Author: Yse good point.

```

> The `"after catch"` is not printed beacuse only the `finally` structure is mirrored in the `Dispose` method. The `catch`, like all other statements between `finally` is not included. statements in between the `catch` and `finally` are not executed in `Dispose`. This may seem odd at first glance but is leaning into the specified behavior for iterator `Dispose`.

Reviewer 3: ```suggestion
The `"after catch"` is not printed beacuse only the `finally` structure is mirrored in the `Dispose` method. The `catch`, like all other statements between `finally` is not included. Statements in between the `catch` and `finally` are not executed in `Dispose`. This may seem odd at first glance but is leaning into the specified behavior for iterator `Dispose`.
```

### Dispose and finally in async iterators

> The `DiposeAsync` behavior for async iterators mirrors that of traditional iterators in that it executes the `finally` blocks at the the point the state machine is suspended. However it does this by setting the `state` variable to a value that represents disposing and then calls `MoveNextAsync`. The implementation of `MoveNextAsync` will then execute only the `finally` blocks for the suspend point. The `DisposeAsync` method does not have a mirror copy of the `finally` blocks.

Reviewer 3: Curious, is there some reason why the codegen of DisposeAsync is different from Dispose?

Author: I suspect it's mostly an accident of history. The `async` state machine didn't have to deal with `Dispose` scenarios initially hence never went through the necessary rewriting process. At the point we added async iterators it was easier to just tweak the `state` field to represent disposing and using the existing state machine. 

@jcouv can confirm or give a better explanation ;)

Reviewer: Yes, it was simpler and more elegant/lightweight to use a single state machine with a "disposal" flag for async-iterators. I was aware of the codegen for iterators but didn't see a benefit to the added overhead (second state machine with bunch of logic extracted into helper methods).

### Code generation of yield inside try / catch in async iterators

> The code generation of async iterators will change such that `catch` blocks do not observably execute during `DisposeAsync`. To achieve this all `catch` blocks visible from a `yield` will rethrow exceptions if the state machine is in a disposing state. For example consider the following code::

Reviewer 3: ```suggestion
The code generation of async iterators will change such that `catch` blocks do not observably execute during `DisposeAsync`. To achieve this all `catch` blocks visible from a `yield` will rethrow exceptions if the state machine is in a disposing state. For example consider the following code:
```

```csharp
var e = M(true).GetEnumerator();
await e.MoveNext();
await e.DisposeAsync();

async IAsyncEnumerable<int> M(bool b)
{
    try
    {
        try
        {
            yield return 13;
            await Task.Yield();
        }
        finally
        {
            Console.WriteLine("inner finally");
            if (b)
            {
                throw new Exception();
            }
        }
    }
    catch
    {
        Console.WriteLine("catch");
    }
    finally
    {
        Console.WriteLine("outer finally");
    }
}
```

This will output:

```cmd
inner finally
outer finally
```

To achieve this the `catch` will be effectively rewritten as follows:

```csharp
catch (Exception ex)
{

>     if (<>1__state == /* dispatching state */)

Reviewer 3: ```suggestion
    if (<>1__state == /* disposing state */)
```

    {
        throw;
    }
    Console.WriteLine("catch");
}
```

The generator will also need to modify any `when` clauses on `catch` blocks to ensure they don't execute during `DisposeAsync`. To do this a prefix will be generated that returns `false` when the state machine is in a disposing state. For example:

```csharp
// User code
catch (Exception ex) when (SomeMethod(ex))

// Generated code

> catch (Exception ex) when (<>1__state == /* dispatching state */ ? false : SomeMethod(ex))

Reviewer 3: ```suggestion
catch (Exception ex) when (<>1__state == /* disposing state */ ? false : SomeMethod(ex))
```

## Considerations

### yield inside finally

Consideration was given to supporting `yield` inside a `finally` block but this creates significant semantic challenges. Consider the following code as an example:

```csharp
var e = M().GetEnumerator();
e.MoveNext();
e.Dispose();

void M()
{
    try
    {
        yield return 42;
    }
    finally
    {
        yield return 13;
    }
}
```

The `finally` block would be executed in `Dispose` which means the language has to decide on the behavior of `yield` during `Dispose`. That could be modeled as:

1. Throw an exception when `yield` is encountered in `Dispose`.
2. Silently ignore the value and continue executing the `finally` block without suspend.

Neither of these seem like desirable outcomes and as such `yield` will not be allowed in `finally` blocks.

## Open Issues

### Preserve catch in Dispose paths

> The `Dispose` method could include mirroring both `catch` and `finally` blocks. This would allow the `catch` block to be executed in `Dispose` when `finally` blocks threw an exceptoin. The approach for this would be to do the followng.

Reviewer 3: ```suggestion
The `Dispose` method could include mirroring both `catch` and `finally` blocks. This would allow the `catch` block to be executed in `Dispose` when `finally` blocks threw an exception. The approach for this would be to do the following.
```

For every `catch` block where the `try` has a nested `try / finally` with `yield`:

1. The contents of the `catch` block will be generated into a parameterless method on the iterator with a `void` return.
2. The contents of the `when` clause will be generated into a parameterless method on the iterator with a `bool` return.
3. The `catch` block will be replaced with a call to the generated method in `MoveNext`.
4. The `when` clause will be replaced with a call to the generated method in `MoveNext`.

> 5. The `Dispose` method will mirror `catch` and `finally` blocks in the same way it mirrors `finally` blocks today and dispatch to the appropriate method

Reviewer 3: Could we also do codegen similar to async iterators, i.e., share MoveNext and Dispose implementations?

Author: That is definitely an option, essentially changing iterators to favor the async iterators code gen approach. I _think_ that would be fine as I can't see any technical reason why iterators have to be done that way but we'd want to vet it out a bit

        }
        finally
        {
            Console.WriteLine("Finally Inner");
        }
    }
    catch (InvalidOperationException ex1) when (ex.Message.Contains("hello"))
    {
        Console.WriteLine("Catch1");
    }
    catch (Exception ex2) 
    {
        Console.WriteLine("Catch2");
    }
    finally
    {
        Console.WriteLine("Finally Outer");
    }
}
static T M<T>(T t) => t;
```

Will generate the following `Dispose` method:

```csharp
[DebuggerHidden]
void IDisposable.Dispose()
{
    int num = <>1__state;
    if ((uint)(num - -4) > 1u && num != 1)
    {
        return;
    }
    try
    {
        if (num != -4 && num != 1)
        {
            return;
        }
        try
        {
        }
        finally
        {
            <>m__Finally2();
        }
    }
    catch (InvalidOperationException ex) when (<>1__ex1 = ex, <>m__When1())
    {
        <>m__Catch1();
    }
    catch (Exception ex)
    {
        <>1__ex2 = ex;
        <>m__Catch1();
    }
    finally
    {
        <>m__Finally1();
    }
}
```

The `<>1__ex1 = ex` in the `when` clause is not legal but the IL generated for the `when` will conceptually have this behavior.

> This would add a bit of complexity the feature and it only produces observable differences when a `finally` block throws an exception on the `Dispose` path. That is likely a rare case. Further it potentially increases the complexity for developers. The `Dispose` method at first glance is likely unintuitive in that it only mirrors `finally` blocks and no other statements but that is also a very simple rule to learn. Preserving `catch` and `finally` and discussing how the code flows between them is potentially more complex.

Reviewer 3: ```suggestion
This would add a bit of complexity to the feature and it only produces observable differences when a `finally` block throws an exception on the `Dispose` path. That is likely a rare case. Further it potentially increases the complexity for developers. The `Dispose` method at first glance is likely unintuitive in that it only mirrors `finally` blocks and no other statements but that is also a very simple rule to learn. Preserving `catch` and `finally` and discussing how the code flows between them is potentially more complex.
```
